File: android/src/main/java/com/emarsys/reactnative/utils/mappers/MessageMapper.kt
package com.emarsys.reactnative.utils.mappers

import com.emarsys.mobileengage.api.action.ActionModel
import com.emarsys.mobileengage.api.action.AppEventActionModel
import com.emarsys.mobileengage.api.action.OpenExternalUrlActionModel
import com.emarsys.mobileengage.api.inbox.Message
import com.emarsys.reactnative.utils.ArrayUtils.toWritableArray
import com.emarsys.reactnative.utils.MapUtils.put
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

object MessageMapper {

  fun toWritableMap(message: Message): WritableMap {
    val writableMap = Arguments.createMap()
    writableMap.put("id", message.id)
    writableMap.put("campaignId", message.campaignId)
    writableMap.put("collapseId", message.collapseId)
    writableMap.put("title", message.title)
    writableMap.put("body", message.body)
    writableMap.put("imageUrl", message.imageUrl)
    writableMap.put("imageAltText", message.imageAltText)
    writableMap.put("receivedAt", message.receivedAt)
    writableMap.put("updatedAt", message.updatedAt)
    writableMap.put("expiresAt", message.expiresAt)
    writableMap.put("tags", message.tags)
    writableMap.put("properties", message.properties)
    writableMap.put("actions", message.actions?.toWritableArray(::toActionModelWritableMap))

    return writableMap
  }

  fun toActionModelWritableMap(actionModel: ActionModel): WritableMap {
    val writableMap = Arguments.createMap()
    writableMap.put("id", actionModel.id)
    writableMap.put("title", actionModel.title)
    writableMap.put("type", actionModel.type)

    when (actionModel) {
      is AppEventActionModel -> {
        writableMap.put("name", actionModel.name)
        writableMap.put("payload", actionModel.payload)
      }
      is OpenExternalUrlActionModel -> {
        writableMap.put("url", actionModel.url)
      }
      else -> {}
    }

    return writableMap
  }
}

File: ios/utils/mappers/MessageMapper.m
#import "MessageMapper.h"

@implementation MessageMapper

static Mapper _toDictionary = ^NSDictionary *(EMSMessage *message) {
  NSMutableDictionary<NSString *, NSObject *> *dictionary = [NSMutableDictionary dictionary];
  dictionary[@"id"] = message.id;
  dictionary[@"campaignId"] = message.campaignId;
  dictionary[@"collapseId"] = message.collapseId;
  dictionary[@"title"] = message.title;
  dictionary[@"body"] = message.body;
  dictionary[@"imageUrl"] = message.imageUrl;
  dictionary[@"imageAltText"] = message.imageAltText;
  dictionary[@"receivedAt"] = message.receivedAt;
  dictionary[@"updatedAt"] = message.updatedAt;
  dictionary[@"expiresAt"] = message.expiresAt;
  dictionary[@"tags"] = message.tags;
  dictionary[@"properties"] = message.properties;
  dictionary[@"actions"] = [message.actions map:MessageMapper.toActionModelDictionary];

  return dictionary;
};

+ (Mapper)toDictionary {
  return _toDictionary;
}

static Mapper _toActionModelDictionary = ^NSDictionary *(id<EMSActionModelProtocol> actionModel) {
  NSMutableDictionary<NSString *, NSObject *> *dictionary = [NSMutableDictionary dictionary];
  dictionary[@"id"] = actionModel.id;
  dictionary[@"title"] = actionModel.title;
  dictionary[@"type"] = actionModel.type;

  if ([actionModel isKindOfClass:[EMSAppEventActionModel class]]) {
    EMSAppEventActionModel *appEventActionModel = (EMSAppEventActionModel *)actionModel;
    dictionary[@"name"] = appEventActionModel.name;
    dictionary[@"payload"] = appEventActionModel.payload;
  } else if ([actionModel isKindOfClass:[EMSOpenExternalUrlActionModel class]]) {
    EMSOpenExternalUrlActionModel *externalUrlActionModel = (EMSOpenExternalUrlActionModel *)actionModel;
    dictionary[@"url"] = externalUrlActionModel.url.absoluteString;
  }

  return dictionary;
};

+ (Mapper)toActionModelDictionary {
  return _toActionModelDictionary;
}

@end
